csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1170,11 +1170,10 @@ private module Stage2 {
   pragma[nomagic]
   private predicate callMayFlowThroughFwd(DataFlowCall call, Configuration config) {
     exists(Ap argAp0, NodeEx out, Cc cc, ApOption argAp, Ap ap |
-      fwdFlow(out, pragma[only_bind_into](cc), pragma[only_bind_into](argAp), ap,
-        pragma[only_bind_into](config)) and
+      fwdFlow(pragma[only_bind_out](out), pragma[only_bind_out](cc), pragma[only_bind_out](argAp),
+        pragma[only_bind_out](ap), pragma[only_bind_out](config)) and
       fwdFlowOutFromArg(call, out, argAp0, ap, config) and
-      fwdFlowIsEntered(call, pragma[only_bind_into](cc), pragma[only_bind_into](argAp), argAp0,
-        pragma[only_bind_into](config))
+      fwdFlowIsEntered(call, cc, argAp, argAp0, config)
```

Comment:
I don't think this rules out the bad join-order - I think it only incidentally improved it. Remember that tagging all the columns of `fwdFlow` with `pragma[only_bind_out]` doesn't really mean much - it only says the `fwdFlow` can't be first in the pipeline, it can still be chosen whenever at least one column is bound, so this bunch of pragmas actually doesn't mean anything.
Given that there's only 3 different join-orders in total (or 6 depending on how you count - they are pairwise semi-isomorphic), and that the set of pragmas you removed were in place to rule out the initial `fwdFlow`-`fwdFlowIsEntered` join, and that you're now seeing an also-bad initial `fwdFlow`-`fwdFlowOutFromArg` that we'll also want to rule out, then we're getting quite close to the point where we could consider `pragma[noopt]`.
Alternatively, I think we'll need the following:
```
      fwdFlow(out, pragma[only_bind_into](cc), pragma[only_bind_into](argAp), ap,
        pragma[only_bind_into](config)) and
      fwdFlowOutFromArg(call, out, argAp0, ap, config) and
      fwdFlowIsEntered(pragma[only_bind_into](call), pragma[only_bind_into](cc), pragma[only_bind_into](argAp), pragma[only_bind_into](argAp0),
        pragma[only_bind_into](config))
```
which forces the order `fwdFlowIsEntered`, `fwdFlowOutFromArg`, `fwdFlow`. (We can also chose `fwdFlowOutFromArg`, `fwdFlowIsEntered`, `fwdFlow` if that seems better).

Reply:
Thanks. Fixed.